### ISSUE TITLE
Trigger pandas.merge after loading two model_station.csv files

### ISSUE THREAD

**Issue by jploveless** [brendanjmeade/result_manager#21]
Referencing the `res_compare` branch https://github.com/brendanjmeade/result_manager/commit/98407c86e3e83d42551c024d29a24ab3d54f7c22

`pandas.merge` seems the easiest way to compare two sets of residual velocities in order to show better/worse model fit. This operation can only be carried out once a station DataFrame is loaded from both folders. My approach so far has been (links point to lines of code in `result_manager.py`:

1. [Create empty DataFrames](https://github.com/brendanjmeade/result_manager/blob/98407c86e3e83d42551c024d29a24ab3d54f7c22/result_manager.py#L153-L154), `station_1` and `station_2` 
2. Within the `load_data` function, it seems like [`station_1`](https://github.com/brendanjmeade/result_manager/blob/98407c86e3e83d42551c024d29a24ab3d54f7c22/result_manager.py#L179) or [`station_2`](https://github.com/brendanjmeade/result_manager/blob/98407c86e3e83d42551c024d29a24ab3d54f7c22/result_manager.py#L185) could get updated through the `station = pd.read_csv('model_station.csv')` command. 
3. So, I placed [some `pandas` code](https://github.com/brendanjmeade/result_manager/blob/98407c86e3e83d42551c024d29a24ab3d54f7c22/result_manager.py#L282-L310) at the bottom of `read_data`, which checks for the existence of the `lon` field in `station_1` and `station_2`. I intend for this to mean that, if both DataFrames have been populated (i.e. `read_data(1)` and `read_data(2)` have been triggered), the `pandas.merge` can take place, creating a new ColumnDataSource, `commonsta`. 
4. `commonsta` is used as the source for [a scatter plotting routine](https://github.com/brendanjmeade/result_manager/blob/98407c86e3e83d42551c024d29a24ab3d54f7c22/result_manager.py#L520-L531), currently set to be always visible so I didn't have to add a checkbox (yet). 

But nothing is appearing, and I'm struggling to understand whether the `station_1` and `station_2` DataFrames are actually getting created within `read_data`, or the regular `station` DataFrame is getting created through `pandas.read_csv` but not getting copied to the suffixed versions. 

@brendanjmeade If you have some time to take a  quick look and share any thoughts/ideas, I'd appreciate it! Regarding point 2 above, the variable assignments seem opposite my intuition; I was expecting to see `stasource_2 = stasource` and not the other way around. But clearly it's working as written, so I tried to follow that template for creating `station_1` and `station_2`, but I think that might be one place my problems lie.

**Comment by brendanjmeade** [comment]
After a quick skim my first thought is that I broke something with the creation of the single loader function.  Perhaps this will be easier to debug without that single function?  I regret writing it!

**Comment by jploveless** [comment]
I now understand the `stasource = stasource_1` style with mutable variables, but because we are reading in `station = pd.read_csv`, we can't follow the same pattern to say `station = station_1`. I think a challenge is a mix of the suffixed ColumnDataSource, which can be defined within `load_data` and then passed to the rest of the code, but `station_1` and `station_2` seem to be within just the scope of the `load_data` function. So I can't figure out whether it's possible for `read_data` to actually compare the `station` DataFrames, because one of them is always out of scope. So, I think it makes sense to move the residual comparison (`pandas.merge`) outside of `read_data`, but now I'm back to square one in asking the title of this issue as a question: how do we trigger the `pandas.merge` only after two station files have been loaded?

**Comment by brendanjmeade** [comment]
This is very clear and helps me to understand this.  What about something like a "compare" button or toggle that calls a function.  If both of the station ColumnDataSources have "real" data in them, then do the comparison?

**Comment by jploveless** [comment]
That's perfect. I was really getting hung up on making things automatic (triggered upon loading the second folder), but a button can be a user-intended way of triggering the conditional logic that checks for existence of both files. This may mean I really need to find a way of doing the station comparison without `pandas`, because it still seems tricky to get DataFrames out of `load_data`. But, I think that's a surmountable problem.

**Comment by brendanjmeade** [comment]
Two very related ideas:

1.  Declare empty dataframes with the appropriate names in the global namespace outside the `load_data` function.  After the `load_data` function populates those dataframes they should be accessible in the global namespace after `load_data` has finished.

2.  If for some reason it's not directly with dataframes there might be a dict intermediary way:
- Declare two empty dicts in the global space before `load_data`.
- Inside `load_data` convert data frame to appropriate dict using the [pandas `to_dict` method](https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_dict.html).  This dictionary should be in the global name space by the time the `load_data` function is done.
- Outside the `load_data` function, you know have dicts with the info you want.  Convert to pandas dataframes with the [pandas `from_dict` method](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.from_dict.html).

Both of these are essentially the same method I used to get the ColumnDataSources out of `load_data` without explicitly returning them.  Maybe?

**Comment by jploveless** [comment]
I do think that DataFrames might be different from the CDSs. I'm pretty sure I had been trying option 1, and it just didn't seem that the `station_1` and `station_2` that I declared as empty dataframes globally were being recognized/processed correctly inside `load_data`. 

I will give the second dict/dataframe approach a try if the method I'm currently working on doesn't end up working. But, I am trying an approach that could just use the `stasource_1` and `stasource_2` CDSs, which we know are being properly parsed out of `load_data`, and finding common stations using [Cantor pairing](https://en.wikipedia.org/wiki/Pairing_function#Cantor_pairing_function) of the lon, lat along with `np.intersect1d`, which returns the indices of common stations for each array. Cantor pairing is designed for combining integers into unique values, but in my tests, it works well with random numbers, so I think it's also working correctly with the lon, lat floats. It helps that I don't need to be a true mapping function, as I don't need to recover the unique lon, lat from the paired values; I just need the indices of the intersected paired values.

**Comment by brendanjmeade** [comment]
What about just converting the `stasource_1` and `stasource_2` to dataframes when needed (e.g., `df = pd.DataFrame(stasource_1.data)`)?

**Comment by jploveless** [comment]
@brendanjmeade bringin that heat

That's brilliant. I think converting to DFs in the "compare" button click function would work really well, then allow the `pandas.merge` approach from before. I'll try to work on this button this weekend. Thank you!

**Comment by jploveless** [comment]
I think this DataFrame-based approach opens up a few other interesting possibilities. What about a new option that shows differences in segment slip rates? We've always used the residual comparison to see where the model _fit_ changes the most, but it might be helpful to see where the model _parameters_ change the most. Definitely a later task, but again, I think enabled by being able to easily compare DFs.

**Comment by brendanjmeade** [comment]
With the client server model, there are python callbacks available as well:

https://docs.bokeh.org/en/latest/docs/user_guide/interaction/python_callbacks.html

Perhaps a big green button isn't needed, but a radio button could do the trick with a python callback that does the diff with the dataframes and then that python callback calls a JavaScript function for display at the end?  Not sure how the last part would work.

**Comment by jploveless** [comment]
Closed with https://github.com/brendanjmeade/result_manager/commit/fdb2e53cd0941b04f2c559f99b0d36122f6a90c6
